pkg/controller/garbagecollector/garbage_collector.go

```diff
@@ -0,0 +1,501 @@
+	clientPool                       dynamic.ClientPool
+	dirtyQueue                       *workqueue.TimedWorkQueue
+	orphanQueue                      *workqueue.TimedWorkQueue
+	monitors                         []monitor
+	propagator                       *Propagator
+	clock                            clock.Clock
+	registeredRateLimiter            *RegisteredRateLimiter
+	registeredRateLimiterForMonitors *RegisteredRateLimiter
+	// GC caches the owners that do not exist according to the API server.
+	absentOwnerCache *UIDCache
+}
+
+func gcListWatcher(client *dynamic.Client, resource schema.GroupVersionResource) *cache.ListWatch {
+	return &cache.ListWatch{
+		ListFunc: func(options v1.ListOptions) (runtime.Object, error) {
+			// APIResource.Kind is not used by the dynamic client, so
+			// leave it empty. We want to list this resource in all
+			// namespaces if it's namespace scoped, so leave
+			// APIResource.Namespaced as false is all right.
+			apiResource := metav1.APIResource{Name: resource.Resource}
+			return client.ParameterCodec(dynamic.VersionedParameterEncoderWithV1Fallback).
+				Resource(&apiResource, v1.NamespaceAll).
+				List(&options)
+		},
+		WatchFunc: func(options v1.ListOptions) (watch.Interface, error) {
+			// APIResource.Kind is not used by the dynamic client, so
+			// leave it empty. We want to list this resource in all
+			// namespaces if it's namespace scoped, so leave
+			// APIResource.Namespaced as false is all right.
+			apiResource := metav1.APIResource{Name: resource.Resource}
+			return client.ParameterCodec(dynamic.VersionedParameterEncoderWithV1Fallback).
+				Resource(&apiResource, v1.NamespaceAll).
+				Watch(&options)
+		},
+	}
+}
+
+func (gc *GarbageCollector) monitorFor(resource schema.GroupVersionResource, kind schema.GroupVersionKind) (monitor, error) {
+	// TODO: consider store in one storage.
+	glog.V(6).Infof("create storage for resource %s", resource)
+	var monitor monitor
+	client, err := gc.metaOnlyClientPool.ClientForGroupVersionKind(kind)
+	if err != nil {
+		return monitor, err
+	}
+	gc.registeredRateLimiterForMonitors.registerIfNotPresent(resource.GroupVersion(), client, "garbage_collector_monitoring")
+	setObjectTypeMeta := func(obj interface{}) {
+		runtimeObject, ok := obj.(runtime.Object)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("expected runtime.Object, got %#v", obj))
+		}
+		runtimeObject.GetObjectKind().SetGroupVersionKind(kind)
+	}
+	monitor.store, monitor.controller = cache.NewInformer(
+		gcListWatcher(client, resource),
+		nil,
+		ResourceResyncTime,
+		cache.ResourceEventHandlerFuncs{
+			// add the event to the propagator's eventQueue.
+			AddFunc: func(obj interface{}) {
+				setObjectTypeMeta(obj)
+				event := &event{
+					eventType: addEvent,
+					obj:       obj,
+				}
+				gc.propagator.eventQueue.Add(&workqueue.TimedWorkQueueItem{StartTime: gc.clock.Now(), Object: event})
+			},
+			UpdateFunc: func(oldObj, newObj interface{}) {
```

Comment:
> Right. Currently the check is done in the propagator.

This is the spot to do it.  Here you have one thread per watch and you can run the checks concurrently.  In the `dependencyGraphBuilder`, you've paid an enqueue, a dequeue, and you're single threaded on the consumption side.